lib/galaxy/datatypes/data.py

```diff
@@ -1205,6 +1205,143 @@ def regex_line_dataprovider(
+        desc="Name of the index file relative to the root folder",
+        readonly=True,
+        optional=False,
+        visible=False,
+    )
+
+    MetadataElement(
+        name="folder_size",
+        default=0,
+        desc="Total size of the folder in bytes",
+        readonly=True,
+        optional=False,
+        visible=False,
+    )
+
+    @classmethod
+    def _get_size(cls, root_path):
+        total_size = 0
+        for dirpath, dirnames, filenames in os.walk(root_path):
+            for f in filenames:
+                fp = os.path.join(dirpath, f)
+                # skip if it is symbolic link
+                if not os.path.islink(fp):
+                    total_size += os.path.getsize(fp)
+        return total_size
+
+    def set_meta(self, dataset: DatasetProtocol, **kwd):
+        efp = dataset.extra_files_path
+        efp_items = os.listdir(efp)
+        root_folder_name = efp_items[0]
+        if len(efp_items) == 1 and os.path.isdir(os.path.join(efp, root_folder_name)):
+            dataset.metadata.root_folder = root_folder_name
+        else:
+            dataset.metadata.root_folder = ""
+        index_file = None
+        for f in self.recognized_index_files:
+            if os.path.isfile(os.path.join(efp, root_folder_name, f)):
+                index_file = f
+                break
+        dataset.metadata.index_file = index_file
+        dataset.metadata.folder_size = self._get_size(os.path.join(efp, dataset.metadata.root_folder))
+
+    def set_peek(self, dataset: DatasetProtocol, **kwd) -> None:
+        if not dataset.dataset.purged:
+            dataset.peek = f"{dataset.metadata.root_folder}"
+            dataset.blurb = nice_size(dataset.metadata.folder_size)
+        else:
+            dataset.peek = "file does not exist"
+            dataset.blurb = "file purged from disk"
+
+    def _archive_main_file(
+        self, archive: ZipstreamWrapper, display_name: str, data_filename: str
+    ) -> Tuple[bool, str, str]:
+        """Overwrites the method to not do anything.
+
+        No main file gets added to a directory archive.
+        """
+        error, msg, messagetype = False, "", ""
+        return error, msg, messagetype
+
+    def display_data(
+        self,
+        trans,
+        dataset: DatasetHasHidProtocol,
+        preview: bool = False,
+        filename: Optional[str] = None,
+        to_ext: Optional[str] = None,
+        **kwd,
+    ):
+        headers = kwd.get("headers", {})
+        # Prevent IE8 from sniffing content type since we're explicit about it.  This prevents intentionally text/plain
+        # content from being rendered in the browser
+        headers["X-Content-Type-Options"] = "nosniff"
+        if to_ext:
+            # Download the directory structure as an archive
+            trans.log_event(f"Download directory for dataset id: {str(dataset.id)}")
+            return self._archive_composite_dataset(trans, dataset, headers, do_action=kwd.get("do_action", "zip"))
```

Comment:
@mvdbeek this is the code that gets repeated to manage the downloading as an archive :)

The rest is for making the eye icon work.

Reply:
Sorry, not following, what do you mean by making the eye icon work ?

Reply:
The index stuff doesn't belong here, directories don't have indexes.

Reply:
yeah, that's the part that will be useful in subclasses (like zarr). That's why the candidate index files list for the parent class is empty. Could be implemented in just relevant subclasses of course.

Reply:
> Sorry, not following, what do you mean by making the eye icon work ?

well, displaying the root folder or the index file. With the current version of directory, when you click the eye icon, you'll just an empty file downloaded.

Reply:
> Could be implemented in just relevant subclasses of course.

But a directory is a relevant, self-contained, out-there-in-the-wild datatype. Let's not add data where none is needed. subclasses should implement this if required.

Reply:
> With the current version of directory, when you click the eye icon, you'll just an empty file downloaded.

Feel free to disable the eye icon on the client or implement a directory browser, but this does not belong into the datatype.